roadmap/implementers-guide/src/runtime-api/README.md

```diff
@@ -0,0 +1,172 @@
+
+```rust
+fn validators() -> Vec<ValidatorId>;
+```
+
+## Validator Groups
+
+Yields the validator groups used during the current session. The validators in the groups are referred to by their index into the validator-set.
+
+```rust
+/// A helper data-type for tracking validator-group rotations.
+struct GroupRotationInfo {
+	session_start_block: BlockNumber,
+	group_rotation_frequency: BlockNumber,
+	now: BlockNumber,
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index,
+	/// assuming the given amount of cores/groups.
+	fn group_for_core(core_index: usize, cores: usize) -> usize;
+}
+
+/// Returns the validator groups and rotation info localized based on the block whose state
+/// this is invoked on. Note that `now` in the `GroupRotationInfo` should be the successor of
+/// the number of the block.
+fn validator_groups() -> (Vec<Vec<ValidatorIndex>>, GroupRotationInfo);
+```
+
+## Availability Cores
+
+Yields information on all availability cores. Cores are either free or occupied. Free cores can have paras assigned to them. Occupied cores don't, but they can become available part-way through a block due to bitfields and then have something scheduled on them. To allow optimistic validation of candidates, the occupied cores are accompanied by information on what is upcoming. This information can be leveraged when validators perceive that there is a high likelihood of a core becoming available based on bitfields seen, and then optimistically validate something that would become scheduled based on that, although there is no guarantee on what the block producer will actually include in the block.
+
+```rust
+fn availability_cores() -> Vec<CoreState>;
+```
+
+This is all the information that a validator needs about scheduling for the current block. It includes all information on [Scheduler](../runtime/scheduler.md) core-assignments and [Inclusion](../runtime/inclusion.md) state of blocks occupying availability cores. It includes data necessary to determine not only which paras are assigned now, but which cores are likely to become freed after processing bitfields, and exactly which bitfields would be necessary to make them so.
+
+```rust
+struct OccupiedCore {
+	/// The ID of the para occupying the core.
+	para: ParaId,
+	/// If this core is freed by availability, this is the assignment that is next up on this
+	/// core, if any. None if there is nothing queued for this core.
+	next_up_on_available: Option<ScheduledCore>,
+	/// The relay-chain block number this began occupying the core at.
+	occupied_since: BlockNumber,
+	/// The relay-chain block this will time-out at, if any.
+	time_out_at: BlockNumber,
+	/// If this core is freed by being timed-out, this is the assignment that is next up on this
+	/// core. None if there is nothing queued for this core or there is no possibility of timing
+	/// out.
+	next_up_on_time_out: Option<ScheduledCore>,
```

Comment:
Having two `next_up_on` fields feels like it introduces the potential for confusion. When would we want `next_up_on_time_out` to differ from `next_up_on_available`? If they do differ, is it possible for a core both to timeout and to become available simultaneously? If so, what's the actual next scheduled core?

Reply:
"the runtime has all the answers".

We don't want them to differ, and yes, it is confusing, but this is purely exposing information about the system described in the scheduler module. It would be nice to have a clear "next up" in all cases, but deeper reading on the scheduler module will reveal why that is not possible.

Reply:
>  is it possible for a core both to timeout and to become available simultaneously

No, they are mutually exclusive. These are the two different paths an `Occupied` core can take to the `Free` state. However, depending on which path is taken (which is unpredictable as it's based on the view/honesty of the block producer), the scheduling metadata can be different, leading to a different assignment onto the now-`Free` core.

Reply:
In general we should always optimize for the `next_up_on_available` case in Node-side code. The reason being that timeouts are only possible at a small subset of blocks, as they are only triggered within the short span of time directly following a group rotation. And even when timeouts can be triggered, unless validators are offline, they will not be reached before availability. And if validators are offline, it's fine to degrade throughput of paras somewhat.

However this runtime API is in the interest of making all information available to the node so more advanced strategies can be taken as research evolves.